File: Python/interval_tree_utils/interval_tree.py
"""
区间树（Interval Tree）实现
================================

一种用于高效处理区间查询的增强二叉搜索树。

时间复杂度：
- 插入：O(log n) 平均
- 删除：O(log n) 平均
- 点查询：O(log n + k)，k 为结果数量
- 区间重叠查询：O(log n + k)，k 为结果数量

空间复杂度：O(n)
"""

from typing import Generic, TypeVar, List, Optional, Iterator, Callable
from dataclasses import dataclass
import random

from .interval import Interval
from .node import IntervalNode

T = TypeVar('T', int, float)


class IntervalTree(Generic[T]):
    """
    区间树
    
    用于高效查询区间重叠的数据结构。
    
    示例:
        >>> tree = IntervalTree()
        >>> tree.insert(Interval(1, 5))
        >>> tree.insert(Interval(3, 8))
        >>> tree.query_point(4)  # 查找包含点 4 的区间
        [Interval(1, 5), Interval(3, 8)]
        >>> tree.query_overlaps(Interval(4, 6))  # 查找与区间 [4,6] 重叠的区间
        [Interval(1, 5), Interval(3, 8)]
    """
    
    def __init__(self, intervals: Optional[List[Interval[T]]] = None):
        """
        初始化区间树
        
        参数:
            intervals: 可选的初始区间列表
        """
        self._root: Optional[IntervalNode[T]] = None
        self._size: int = 0
        
        if intervals:
            for interval in intervals:
                self.insert(interval)
    
    def insert(self, interval: Interval[T]) -> None:
        """
        插入一个区间
        
        参数:
            interval: 要插入的区间
        """
        if self._root is None:
            center = self._compute_center([interval])
            self._root = IntervalNode(center=center)
            self._root.add_interval(interval)
        else:
            self._insert_node(self._root, interval)
        self._size += 1
    
    def _compute_center(self, intervals: List[Interval[T]]) -> T:
        """计算区间的中心点"""
        if not intervals:
            return 0
        starts = [i.start for i in intervals]
        ends = [i.end for i in intervals]
        center = (min(starts) + max(ends)) / 2.0
        # 如果输入是整数，返回整数中心
        if all(isinstance(i.start, int) and isinstance(i.end, int) for i in intervals):
            return int(center)
        return center
    
    def _insert_node(self, node: IntervalNode[T], interval: Interval[T]) -> None:
        """递归插入区间到合适的位置"""
        # 区间与中心点重叠
        if interval.start <= node.center <= interval.end:
            node.add_interval(interval)
            return
        
        # 区间完全在中心点左侧
        if interval.end < node.center:
            if node.left is None:
                node.left = IntervalNode(center=self._compute_center([interval]))
                node.left.add_interval(interval)
                node.left.update_bounds()
            else:
                self._insert_node(node.left, interval)
            node.update_bounds()
            return
        
        # 区间完全在中心点右侧
        if interval.start > node.center:
            if node.right is None:
                node.right = IntervalNode(center=self._compute_center([interval]))
                node.right.add_interval(interval)
                node.right.update_bounds()
            else:
                self._insert_node(node.right, interval)
            node.update_bounds()
            return
    
    def remove(self, interval: Interval[T]) -> bool:
        """
        移除一个区间
        
        参数:
            interval: 要移除的区间
            
        返回:
            是否成功移除
        """
        if self._root is None:
            return False
        
        result = self._remove_node(self._root, interval)
        if result:
            self._size -= 1
        return result
    
    def _remove_node(self, node: IntervalNode[T], interval: Interval[T]) -> bool:
        """递归移除区间"""
        # 检查当前节点的重叠区间列表
        if interval in node.overlapping:
            node.remove_interval(interval)
            # 如果节点变空且有子节点，可能需要重建
            if not node.overlapping and node.is_leaf():
                # 叶节点且无区间，可以安全删除
                pass  # 简单处理，保留空节点
            return True
        
        # 递归搜索
        if interval.end < node.center and node.left:
            return self._remove_node(node.left, interval)
        elif interval.start > node.center and node.right:
            return self._remove_node(node.right, interval)
        else:
            # 需要搜索两边
            found = False
            if node.left:
                found = self._remove_node(node.left, interval)
            if not found and node.right:
                found = self._remove_node(node.right, interval)
            return found
    
    def query_point(self, point: T) -> List[Interval[T]]:
        """
        查询包含指定点的所有区间
        
        参数:
            point: 查询点
            
        返回:
            包含该点的区间列表
        """
        if self._root is None:
            return []
        
        result: List[Interval[T]] = []
        self._query_point_node(self._root, point, result)
        return result
    
    def _query_point_node(self, node: IntervalNode[T], point: T, result: List[Interval[T]]) -> None:
        """递归查询包含指定点的区间"""
        # 检查当前节点的区间
        for interval in node.overlapping:
            if interval.contains_point(point):
                result.append(interval)
        
        # 检查是否需要搜索左子树
        if node.left and point <= node.center:
            self._query_point_node(node.left, point, result)
        
        # 检查是否需要搜索右子树
        if node.right and point >= node.center:
            self._query_point_node(node.right, point, result)
    
    def query_overlaps(self, interval: Interval[T]) -> List[Interval[T]]:
        """
        查询与指定区间重叠的所有区间
        
        参数:
            interval: 查询区间
            
        返回:
            与该区间重叠的区间列表
        """
        if self._root is None:
            return []
        
        result: List[Interval[T]] = []
        self._query_overlaps_node(self._root, interval, result)
        return result
    
    def _query_overlaps_node(self, node: IntervalNode[T], interval: Interval[T], result: List[Interval[T]]) -> None:
        """递归查询重叠区间"""
        # 检查当前节点的区间
        for iv in node.overlapping:
            if iv.overlaps(interval):
                result.append(iv)
        
        # 检查是否需要搜索左子树
        if node.left and interval.start <= node.center:
            self._query_overlaps_node(node.left, interval, result)
        
        # 检查是否需要搜索右子树
        if node.right and interval.end >= node.center:
            self._query_overlaps_node(node.right, interval, result)
    
    def query_contains(self, interval: Interval[T]) -> List[Interval[T]]:
        """
        查询完全包含指定区间的区间
        
        参数:
            interval: 查询区间
            
        返回:
            完全包含该区间的区间列表
        """
        if self._root is None:
            return []
        
        result: List[Interval[T]] = []
        self._query_contains_node(self._root, interval, result)
        return result
    
    def _query_contains_node(self, node: IntervalNode[T], interval: Interval[T], result: List[Interval[T]]) -> None:
        """递归查询包含指定区间的区间"""
        for iv in node.overlapping:
            if iv.contains(interval):
                result.append(iv)
        
        # 搜索左子树（包含区间可能在左边）
        if node.left and node.left.max_end is not None and node.left.max_end >= interval.end:
            self._query_contains_node(node.left, interval, result)
        
        # 搜索右子树
        if node.right and node.right.min_start is not None and node.right.min_start <= interval.start:
            self._query_contains_node(node.right, interval, result)
    
    def query_contained_by(self, interval: Interval[T]) -> List[Interval[T]]:
        """
        查询被指定区间完全包含的区间
        
        参数:
            interval: 查询区间
            
        返回:
            被该区间完全包含的区间列表
        """
        if self._root is None:
            return []
        
        result: List[Interval[T]] = []
        self._query_contained_by_node(self._root, interval, result)
        return result
    
    def _query_contained_by_node(self, node: IntervalNode[T], interval: Interval[T], result: List[Interval[T]]) -> None:
        """递归查询被指定区间包含的区间"""
        for iv in node.overlapping:
            if interval.contains(iv):
                result.append(iv)
        
        # 搜索左子树
        if node.left and node.left.min_start is not None and node.left.min_start >= interval.start:
            self._query_contained_by_node(node.left, interval, result)
        
        # 搜索右子树
        if node.right and node.right.max_end is not None and node.right.max_end <= interval.end:
            self._query_contained_by_node(node.right, interval, result)
    
    def find_first_overlapping(self, interval: Interval[T]) -> Optional[Interval[T]]:
        """查找第一个与指定区间重叠的区间"""
        results = self.query_overlaps(interval)
        return results[0] if results else None
    
    def find_all_gaps(self, start: T, end: T) -> List[Interval[T]]:
        """
        找出指定范围内未被任何区间覆盖的空白区域
        
        参数:
            start: 范围起点
            end: 范围终点
            
        返回:
            空白区间列表
        """
        # 收集范围内所有区间并排序
        all_intervals = sorted(self.query_overlaps(Interval(start, end)), key=lambda x: x.start)
        
        gaps: List[Interval[T]] = []
        current = start
        
        for interval in all_intervals:
            if interval.start > current:
                gaps.append(Interval(current, interval.start - 1))
            current = max(current, interval.end + 1)
        
        if current <= end:
            gaps.append(Interval(current, end))
        
        return gaps
    
    def __len__(self) -> int:
        """返回区间数量"""
        return self._size
    
    def __bool__(self) -> bool:
        """检查树是否非空"""
        return self._size > 0
    
    def __contains__(self, interval: Interval[T]) -> bool:
        """检查区间是否在树中"""
        results = self.query_overlaps(interval)
        return interval in results
    
    def __iter__(self) -> Iterator[Interval[T]]:
        """迭代所有区间"""
        if self._root:
            yield from self._root.collect_all()
    
    def clear(self) -> None:
        """清空区间树"""
        self._root = None
        self._size = 0
    
    def height(self) -> int:
        """返回树的高度"""
        return self._root.height() if self._root else 0
    
    def is_empty(self) -> bool:
        """检查树是否为空"""
        return self._size == 0
    
    def to_list(self) -> List[Interval[T]]:
        """将所有区间转为列表"""
        return list(self) if self._root else []
    
    def get_statistics(self) -> dict:
        """获取树的统计信息"""
        return {
            'size': self._size,
            'height': self.height(),
            'is_empty': self.is_empty(),
            'node_count': self._root.size() if self._root else 0
        }
    
    def balance(self) -> None:
        """重建平衡的区间树"""
        if self._size <= 1:
            return
        
        intervals = self.to_list()
        self.clear()
        
        # 使用中位数分割策略重建
        self._build_balanced(intervals)
    
    def _build_balanced(self, intervals: List[Interval[T]]) -> None:
        """使用中位数分割构建平衡树"""
        if not intervals:
            return
        
        intervals.sort()
        self._build_balanced_recursive(intervals)
    
    def _build_balanced_recursive(self, intervals: List[Interval[T]]) -> None:
        """递归构建平衡树"""
        if not intervals:
            return
        
        mid = len(intervals) // 2
        center_interval = intervals[mid]
        
        # 收集与中心区间重叠的所有区间
        overlapping = [center_interval]
        left_intervals = []
        right_intervals = []
        
        for interval in intervals:
            if interval == center_interval:
                continue
            if interval.end < center_interval.start:
                left_intervals.append(interval)
            elif interval.start > center_interval.end:
                right_intervals.append(interval)
            else:
                overlapping.append(interval)
        
        # 插入重叠区间
        for interval in overlapping:
            self.insert(interval)
        
        # 递归处理左右区间
        self._build_balanced_recursive(left_intervals)
        self._build_balanced_recursive(right_intervals)
    
    @classmethod
    def from_tuples(cls, intervals: List[tuple]) -> 'IntervalTree[T]':
        """
        从元组列表创建区间树
        
        参数:
            intervals: 元组列表，每个元组为 (start, end) 或 (start, end, value)
            
        返回:
            新的区间树实例
        """
        tree = cls()
        for t in intervals:
            if len(t) == 2:
                tree.insert(Interval(t[0], t[1]))
            else:
                tree.insert(Interval(t[0], t[1], t[2]))
        return tree
    
    def __repr__(self) -> str:
        return f"IntervalTree(size={self._size}, height={self.height()})"

File: Go/interval_tree_utils/interval_tree.go
// Package interval_tree_utils provides an Interval Tree (augmented Binary Search Tree)
// for efficient overlapping interval queries.
//
// An interval tree stores intervals and allows querying for all intervals that overlap
// with a given point or interval. Time complexity for insertion and deletion is O(log n),
// and for querying O(log n + k) where k is the number of results.
package interval_tree_utils

import (
	"fmt"
	"sort"
)

// Interval represents a range with a start and end point.
// The interval is half-open: [Start, End) - includes Start but not End.
type Interval struct {
	Start int
	End   int
	Data  interface{} // Optional user data associated with the interval
}

// NewInterval creates a new interval with optional data.
func NewInterval(start, end int, data interface{}) Interval {
	return Interval{Start: start, End: end, Data: data}
}

// Contains checks if a point is within the interval.
func (i Interval) Contains(point int) bool {
	return point >= i.Start && point < i.End
}

// Overlaps checks if this interval overlaps with another interval.
func (i Interval) Overlaps(other Interval) bool {
	return i.Start < other.End && other.Start < i.End
}

// Length returns the length of the interval.
func (i Interval) Length() int {
	return i.End - i.Start
}

// String returns a string representation of the interval.
func (i Interval) String() string {
	if i.Data != nil {
		return fmt.Sprintf("[%d, %d) %v", i.Start, i.End, i.Data)
	}
	return fmt.Sprintf("[%d, %d)", i.Start, i.End)
}

// Node represents a node in the interval tree.
type Node struct {
	Interval    Interval
	MaxEnd      int  // Maximum end value in the subtree rooted at this node
	Left        *Node
	Right       *Node
	Height      int  // Height of the node for AVL balancing
	Balance     int  // Balance factor for AVL
}

// IntervalTree is an AVL-balanced interval tree.
type IntervalTree struct {
	Root   *Node
	Count  int
}

// NewIntervalTree creates a new empty interval tree.
func NewIntervalTree() *IntervalTree {
	return &IntervalTree{}
}

// Insert adds an interval to the tree.
func (t *IntervalTree) Insert(interval Interval) {
	t.Root = t.insert(t.Root, interval)
	t.Count++
}

func (t *IntervalTree) insert(node *Node, interval Interval) *Node {
	if node == nil {
		return &Node{
			Interval: interval,
			MaxEnd:  interval.End,
			Height:  1,
		}
	}

	if interval.Start < node.Interval.Start {
		node.Left = t.insert(node.Left, interval)
	} else {
		node.Right = t.insert(node.Right, interval)
	}

	// Update max end
	node.MaxEnd = max(node.Interval.End, max(getMaxEnd(node.Left), getMaxEnd(node.Right)))

	// Update height and balance
	node.Height = 1 + max(getHeight(node.Left), getHeight(node.Right))
	node.Balance = getHeight(node.Left) - getHeight(node.Right)

	// Rebalance if needed (AVL)
	return t.rebalance(node)
}

// Delete removes an interval from the tree.
func (t *IntervalTree) Delete(interval Interval) bool {
	if t.Root == nil {
		return false
	}
	
	found := false
	t.Root, found = t.delete(t.Root, interval)
	if found {
		t.Count--
	}
	return found
}

func (t *IntervalTree) delete(node *Node, interval Interval) (*Node, bool) {
	if node == nil {
		return nil, false
	}

	found := false

	if interval.Start < node.Interval.Start {
		node.Left, found = t.delete(node.Left, interval)
	} else if interval.Start > node.Interval.Start {
		node.Right, found = t.delete(node.Right, interval)
	} else {
		// Found potential match - check if same interval
		if node.Interval.End == interval.End && node.Interval.Data == interval.Data {
			found = true
			if node.Left == nil {
				return node.Right, true
			}
			if node.Right == nil {
				return node.Left, true
			}
			// Node has two children - find inorder successor
			successor := t.findMin(node.Right)
			node.Interval = successor.Interval
			node.Right, _ = t.delete(node.Right, successor.Interval)
		} else {
			// Same start but different - search both sides
			node.Left, found = t.delete(node.Left, interval)
			if !found {
				node.Right, found = t.delete(node.Right, interval)
			}
		}
	}

	if node == nil {
		return nil, found
	}

	// Update max end
	node.MaxEnd = max(node.Interval.End, max(getMaxEnd(node.Left), getMaxEnd(node.Right)))

	// Update height and balance
	node.Height = 1 + max(getHeight(node.Left), getHeight(node.Right))
	node.Balance = getHeight(node.Left) - getHeight(node.Right)

	return t.rebalance(node), found
}

func (t *IntervalTree) findMin(node *Node) *Node {
	for node.Left != nil {
		node = node.Left
	}
	return node
}

// QueryPoint finds all intervals that contain the given point.
func (t *IntervalTree) QueryPoint(point int) []Interval {
	var results []Interval
	t.queryPoint(t.Root, point, &results)
	return results
}

func (t *IntervalTree) queryPoint(node *Node, point int, results *[]Interval) {
	if node == nil {
		return
	}

	// If point is greater than max end in this subtree, no overlap possible
	if point > node.MaxEnd {
		return
	}

	// Check left subtree
	t.queryPoint(node.Left, point, results)

	// Check current node
	if node.Interval.Contains(point) {
		*results = append(*results, node.Interval)
	}

	// If point is less than interval start, no need to check right
	if point < node.Interval.Start {
		return
	}

	// Check right subtree
	t.queryPoint(node.Right, point, results)
}

// QueryOverlaps finds all intervals that overlap with the given interval.
func (t *IntervalTree) QueryOverlaps(interval Interval) []Interval {
	var results []Interval
	t.queryOverlaps(t.Root, interval, &results)
	return results
}

func (t *IntervalTree) queryOverlaps(node *Node, interval Interval, results *[]Interval) {
	if node == nil {
		return
	}

	// If interval starts after max end in subtree, no overlap possible
	if interval.Start >= node.MaxEnd {
		return
	}

	// Check left subtree
	t.queryOverlaps(node.Left, interval, results)

	// Check current node
	if node.Interval.Overlaps(interval) {
		*results = append(*results, node.Interval)
	}

	// If interval ends before current interval starts, no need to check right
	if interval.End <= node.Interval.Start {
		return
	}

	// Check right subtree
	t.queryOverlaps(node.Right, interval, results)
}

// ContainsPoint checks if any interval contains the given point.
func (t *IntervalTree) ContainsPoint(point int) bool {
	return t.containsPoint(t.Root, point)
}

func (t *IntervalTree) containsPoint(node *Node, point int) bool {
	if node == nil {
		return false
	}

	if point > node.MaxEnd {
		return false
	}

	if node.Interval.Contains(point) {
		return true
	}

	if point < node.Interval.Start {
		return t.containsPoint(node.Left, point)
	}

	return t.containsPoint(node.Left, point) || t.containsPoint(node.Right, point)
}

// HasOverlap checks if any interval overlaps with the given interval.
func (t *IntervalTree) HasOverlap(interval Interval) bool {
	return t.hasOverlap(t.Root, interval)
}

func (t *IntervalTree) hasOverlap(node *Node, interval Interval) bool {
	if node == nil {
		return false
	}

	if interval.Start >= node.MaxEnd {
		return false
	}

	if node.Interval.Overlaps(interval) {
		return true
	}

	if interval.End <= node.Interval.Start {
		return t.hasOverlap(node.Left, interval)
	}

	return t.hasOverlap(node.Left, interval) || t.hasOverlap(node.Right, interval)
}

// GetAll returns all intervals in the tree.
func (t *IntervalTree) GetAll() []Interval {
	var results []Interval
	t.inorder(t.Root, &results)
	return results
}

func (t *IntervalTree) inorder(node *Node, results *[]Interval) {
	if node == nil {
		return
	}
	t.inorder(node.Left, results)
	*results = append(*results, node.Interval)
	t.inorder(node.Right, results)
}

// IsEmpty checks if the tree is empty.
func (t *IntervalTree) IsEmpty() bool {
	return t.Root == nil
}

// Size returns the number of intervals in the tree.
func (t *IntervalTree) Size() int {
	return t.Count
}

// Clear removes all intervals from the tree.
func (t *IntervalTree) Clear() {
	t.Root = nil
	t.Count = 0
}

// Height returns the height of the tree.
func (t *IntervalTree) Height() int {
	return getHeight(t.Root)
}

// FindGaps returns gaps between intervals that don't overlap with any interval in the tree.
// The range parameter specifies the overall range to search for gaps.
func (t *IntervalTree) FindGaps(rangeStart, rangeEnd int) []Interval {
	allIntervals := t.GetAll()
	if len(allIntervals) == 0 {
		return []Interval{{Start: rangeStart, End: rangeEnd}}
	}

	// Sort by start
	sort.Slice(allIntervals, func(i, j int) bool {
		return allIntervals[i].Start < allIntervals[j].Start
	})

	var gaps []Interval
	currentEnd := rangeStart

	for _, iv := range allIntervals {
		if iv.Start > currentEnd && iv.Start > rangeStart {
			gapStart := max(currentEnd, rangeStart)
			gapEnd := min(iv.Start, rangeEnd)
			if gapStart < gapEnd {
				gaps = append(gaps, Interval{Start: gapStart, End: gapEnd})
			}
		}
		if iv.End > currentEnd {
			currentEnd = iv.End
		}
	}

	// Check for gap at the end
	if currentEnd < rangeEnd {
		gaps = append(gaps, Interval{Start: max(currentEnd, rangeStart), End: rangeEnd})
	}

	return gaps
}

// MergeOverlapping returns a new slice of intervals with overlapping intervals merged.
func MergeOverlapping(intervals []Interval) []Interval {
	if len(intervals) == 0 {
		return intervals
	}

	// Sort by start
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i].Start < intervals[j].Start
	})

	var merged []Interval
	current := intervals[0]

	for i := 1; i < len(intervals); i++ {
		if intervals[i].Start <= current.End {
			// Overlapping - merge
			if intervals[i].End > current.End {
				current.End = intervals[i].End
			}
		} else {
			merged = append(merged, current)
			current = intervals[i]
		}
	}
	merged = append(merged, current)

	return merged
}

// Coverage calculates total coverage (union length) of intervals.
func Coverage(intervals []Interval) int {
	merged := MergeOverlapping(intervals)
	total := 0
	for _, iv := range merged {
		total += iv.Length()
	}
	return total
}

// Intersection returns the intersection of two intervals, or false if they don't overlap.
func Intersection(a, b Interval) (Interval, bool) {
	if !a.Overlaps(b) {
		return Interval{}, false
	}
	return Interval{
		Start: max(a.Start, b.Start),
		End:   min(a.End, b.End),
	}, true
}

// Union returns the union of two intervals if they overlap or are adjacent.
func Union(a, b Interval) (Interval, bool) {
	if !a.Overlaps(b) && a.End != b.Start && b.End != a.Start {
		return Interval{}, false
	}
	return Interval{
		Start: min(a.Start, b.Start),
		End:   max(a.End, b.End),
	}, true
}

// Helper functions

func getHeight(node *Node) int {
	if node == nil {
		return 0
	}
	return node.Height
}

func getMaxEnd(node *Node) int {
	if node == nil {
		return 0
	}
	return node.MaxEnd
}

func (t *IntervalTree) rebalance(node *Node) *Node {
	// Left heavy
	if node.Balance > 1 {
		if getHeight(node.Left.Left) >= getHeight(node.Left.Right) {
			return t.rotateRight(node)
		}
		node.Left = t.rotateLeft(node.Left)
		return t.rotateRight(node)
	}

	// Right heavy
	if node.Balance < -1 {
		if getHeight(node.Right.Right) >= getHeight(node.Right.Left) {
			return t.rotateLeft(node)
		}
		node.Right = t.rotateRight(node.Right)
		return t.rotateLeft(node)
	}

	return node
}

func (t *IntervalTree) rotateRight(y *Node) *Node {
	x := y.Left
	T2 := x.Right

	x.Right = y
	y.Left = T2

	// Update heights
	y.Height = 1 + max(getHeight(y.Left), getHeight(y.Right))
	x.Height = 1 + max(getHeight(x.Left), getHeight(x.Right))

	// Update balance factors
	y.Balance = getHeight(y.Left) - getHeight(y.Right)
	x.Balance = getHeight(x.Left) - getHeight(x.Right)

	// Update max ends
	y.MaxEnd = max(y.Interval.End, max(getMaxEnd(y.Left), getMaxEnd(y.Right)))
	x.MaxEnd = max(x.Interval.End, max(getMaxEnd(x.Left), getMaxEnd(x.Right)))

	return x
}

func (t *IntervalTree) rotateLeft(x *Node) *Node {
	y := x.Right
	T2 := y.Left

	y.Left = x
	x.Right = T2

	// Update heights
	x.Height = 1 + max(getHeight(x.Left), getHeight(x.Right))
	y.Height = 1 + max(getHeight(y.Left), getHeight(y.Right))

	// Update balance factors
	x.Balance = getHeight(x.Left) - getHeight(x.Right)
	y.Balance = getHeight(y.Left) - getHeight(y.Right)

	// Update max ends
	x.MaxEnd = max(x.Interval.End, max(getMaxEnd(x.Left), getMaxEnd(x.Right)))
	y.MaxEnd = max(y.Interval.End, max(getMaxEnd(y.Left), getMaxEnd(y.Right)))

	return y
}

// String returns a visual representation of the tree (for debugging).
func (t *IntervalTree) String() string {
	if t.Root == nil {
		return "Empty tree"
	}
	return t.stringHelper(t.Root, "", true)
}

func (t *IntervalTree) stringHelper(node *Node, prefix string, isTail bool) string {
	if node == nil {
		return ""
	}

	var result string
	result += t.stringHelper(node.Right, prefix+"│   ", false)
	result += fmt.Sprintf("%s%s [%d,%d) maxEnd=%d\n", prefix, getBranch(isTail), node.Interval.Start, node.Interval.End, node.MaxEnd)
	result += t.stringHelper(node.Left, prefix+"│   ", true)

	return result
}

func getBranch(isTail bool) string {
	if isTail {
		return "└── "
	}
	return "┌── "
}